libsolidity/analysis/TypeChecker.cpp

```diff
@@ -2445,6 +2451,131 @@ void TypeChecker::typeCheckABIEncodeCallFunction(FunctionCall const& _functionCa
 	}
 }
 
+void TypeChecker::typeCheckABIEncodeErrorFunction(FunctionCall const& _functionCall)
+{
+	std::vector<ASTPointer<Expression const>> const& arguments = _functionCall.arguments();
+
+	// Expecting first argument to be the function pointer and second to be a tuple.
+	if (arguments.size() != 2)
+	{
+		m_errorReporter.typeError(
+			6220_error,
+			_functionCall.location(),
+			"Expected two arguments: a custom error followed by a tuple."
+		);
+		return;
+	}
+
+	FunctionType const* externalFunctionType = nullptr;
+	if (auto const functionPointerType = dynamic_cast<FunctionTypePointer>(type(*arguments.front())))
+	{
+		// this cannot be a library function, that is checked below
+		externalFunctionType = functionPointerType->asExternallyCallableFunction(false);
+		solAssert(externalFunctionType->kind() == functionPointerType->kind());
+	}
+	else
+	{
+		m_errorReporter.typeError(
+			5512_error,
+			arguments.front()->location(),
+			"Expected first argument to be a custom error, not \"" +
+			type(*arguments.front())->humanReadableName() +
+			"\"."
+		);
+		return;
+	}
+
+	if (externalFunctionType->kind() != FunctionType::Kind::Error)
+	{
+		std::string msg = "Expected an error type.";
+
+		switch (externalFunctionType->kind())
+		{
+			case FunctionType::Kind::Internal:
+			case FunctionType::Kind::External:
+			case FunctionType::Kind::Declaration:
+			case FunctionType::Kind::DelegateCall:
+				msg += " Cannot use functions for abi.encodeError.";
+				break;
+			case FunctionType::Kind::Creation:
+				msg += " Provided creation function.";
+				break;
+			case FunctionType::Kind::Event:
+				msg += " Cannot use events for abi.encodeError.";
+				break;
+			default:
+				msg += " Cannot use special function.";
+		}
+
+		m_errorReporter.typeError(3510_error, arguments[0]->location(), {}, msg);
```

Comment:
```suggestion
		m_errorReporter.typeError(3510_error, arguments[0]->location(), msg);
```
We also have one overload which accepts just one location.